components/nexus-core/src/test/java/org/sonatype/nexus/proxy/repository/AnAbstractProxyRepositoryRetrieveRemoteItemMethodTest.java

```diff
@@ -0,0 +1,225 @@
+    {
+      @Override
+      public RepositoryItemUid answer(final InvocationOnMock invocationOnMock) throws Throwable {
+        return repositoryItemUidFactory.createUid(abstractRepository, (String) invocationOnMock.getArguments()[0]);
+      }
+    });
+
+    this.file = new DefaultStorageFileItem(abstractRepository, new ResourceStoreRequest("/foo"), true, true,
+        new StringContentLocator("irrelevant"));
+    this.coll = new DefaultStorageCollectionItem(abstractRepository, new ResourceStoreRequest("/foo"), true, true);
+  }
+
+  @Test
+  public void fileRequestedFoundRemotelyButFailsCaching() throws Exception {
+    when(remoteRepositoryStorage
+        .retrieveItem(any(ProxyRepository.class), any(ResourceStoreRequest.class), anyString())).thenReturn(
+        file);
+    doThrow(new LocalStorageException("unimportant")).when(localRepositoryStorage).storeItem(any(Repository.class),
+        any(StorageItem.class));
+    when(localRepositoryStorage.retrieveItem(any(Repository.class), any(ResourceStoreRequest.class))).thenReturn(file);
+    try {
+      abstractRepository.doRetrieveRemoteItem(new ResourceStoreRequest("/foo"));
+    } catch (ItemNotFoundException e) {
+      // we expect this
+    }
+    // we need cleanup: file was returned and caching was attempted
+    verify(localRepositoryStorage, times(1)).deleteItem(any(Repository.class), any(ResourceStoreRequest.class));
+  }
+
+  @Test
+  public void directoryRequestedFoundRemotelyButFailsCaching() throws Exception {
+    when(remoteRepositoryStorage.retrieveItem(any(ProxyRepository.class), any(ResourceStoreRequest.class), anyString())).thenReturn(
+        file);
+    doThrow(new LocalStorageException("unimportant")).when(localRepositoryStorage).storeItem(any(Repository.class),
+        any(StorageItem.class));
+    when(localRepositoryStorage.retrieveItem(any(Repository.class), any(ResourceStoreRequest.class))).thenReturn(coll);
+    try {
+      abstractRepository.doRetrieveRemoteItem(new ResourceStoreRequest("/foo"));
+    } catch (ItemNotFoundException e) {
+      // we expect this
+    }
+    // we do not need cleanup: file was returned and caching was attempted but LS has coll
+    verify(localRepositoryStorage, times(0)).deleteItem(any(Repository.class), any(ResourceStoreRequest.class));
+  }
+
+  @Test
+  public void fileRequestedNotFoundRemotely() throws Exception {
+    // remote throws RemoteItemNotFoundEx
+    final RemoteItemNotFoundException rinf = new RemoteItemNotFoundException(new ResourceStoreRequest("/foo"), abstractRepository, "nan", "nan");
+    when(remoteRepositoryStorage.retrieveItem(any(ProxyRepository.class), any(ResourceStoreRequest.class), anyString()))
+        .thenThrow(rinf);
+
+    when(localRepositoryStorage.retrieveItem(any(Repository.class), any(ResourceStoreRequest.class))).thenReturn(file);
+    try {
+      abstractRepository.doRetrieveRemoteItem(new ResourceStoreRequest("/foo"));
```

Comment:
ResourceStoreRequest can similarly be extracted for reuse rather than defined per test case